### PR TITLE
Preliminary shared work

### PR DESCRIPTION
**Pull request by UplinkCoder**
This will block read/write access to shared variables
execpt for the casting shared away operation, and when passing a shared argument to a shared parameter.

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -1288,7 +1288,7 @@ private Expression resolvePropertiesX(Scope* sc, Expression e1, Expression e2 =
     {
         VarExp ve = cast(VarExp)e1;
         VarDeclaration v = ve.var.isVarDeclaration();
-        if (v && ve.checkPurity(sc, v))
+        if (v && (ve.checkPurity(sc, v) || checkAccsessShared(ve.loc, v)))
```
Comment (Geod24) [inline src/dmd/expressionsem.d]: `Access` not `Accsess`

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -2376,6 +2400,8 @@ private Module loadStdMath()
     return impStdMath.mod;
 }
 
+bool shared_allowed = false;
```
Comment (Geod24) [inline src/dmd/expressionsem.d]: Don't use globals

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -4013,6 +4039,41 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
     {
+        const old_shared_allowed = shared_allowed;
+        scope(exit)
+            shared_allowed = old_shared_allowed;
+
+        // @@@SHARED@@@
+        // we need to set allow shared for overload resolution
+        shared_allowed = true;
+        // after resolution is done
+        scope (exit)
```
Comment (Geod24) [inline src/dmd/expressionsem.d]: That's way too much code to put in a `scope (exit)`

File: src/dmd/expressionsem.d [hunk 4]
```diff
@@ -5496,6 +5557,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
     override void visit(BinAssignExp exp)
     {
+
```
Comment (Geod24) [inline src/dmd/expressionsem.d]: Unrelated diff

File: src/dmd/expressionsem.d (hunk 2 above)
Comment (UplinkCoder) [reply src/dmd/expressionsem.d]: there's no way around that would not involve passing a state pointer to every name-resolving function in dmd

File: src/dmd/expressionsem.d (hunk 4 above)
Comment (UplinkCoder) [reply src/dmd/expressionsem.d]: true fixing

File: src/dmd/expressionsem.d (hunk 3 above)
Comment (UplinkCoder) [reply src/dmd/expressionsem.d]: I have to ensure this runs before every return from here.
maybe later it can be cleaner.
I am under time-pressure.

File: src/dmd/expressionsem.d (hunk 2 above)
Comment (UplinkCoder) [reply src/dmd/expressionsem.d]: I can put it in a struct and thereby making the global access easier to spot.

File: src/dmd/expressionsem.d [hunk 5]
```diff
@@ -6767,6 +6829,9 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
     override void visit(CastExp exp)
     {
+        bool parent_shared_allowed = GlobalStateExpressionSemantic.shared_allowed;
+        scope (exit) GlobalStateExpressionSemantic.shared_allowed = parent_shared_allowed;
```
Comment (ibuclaw) [inline src/dmd/expressionsem.d]: Seems like you should be adding a SCOPE flag, instead of global state.

Comment (UplinkCoder) [reply src/dmd/expressionsem.d]: Thanks! I don't think I've ever used it. How does it work?

Comment (thewilsonator) [reply src/dmd/expressionsem.d]: Add a flag to https://github.com/dlang/dmd/blob/master/src/dmd/dscope.d#L44 and check/set it from the scope parameter of the expression.

Comment (UplinkCoder) [reply src/dmd/expressionsem.d]: This is more granular than a scope.

Comment (thewilsonator) [reply src/dmd/expressionsem.d]: But it is still a recursive analysis, it doesn't matter where you put it.

File: test/fail_compilation/fail_shared.d [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+{
+    x += 7; // Error: Trying to accsess shared state x
+    x = 22; // Error: Trying to accsess shared state x
+    return x; // Error: Trying to accsess shared state x
+
+}
+
+int sync(shared int *x, int y)
+{
+    return x = y; // Error: Trying to accsess shared state x
```
Comment (TurkeyMan) [inline test/fail_compilation/fail_shared.d]: `x` is a pointer... assigning int to pointer?

Comment (UplinkCoder) [reply test/fail_compilation/fail_shared.d]: auto deref.

Comment (TurkeyMan) [reply test/fail_compilation/fail_shared.d]: auto deref on assignment? is that really a thing?

**Comment by TurkeyMan** [comment]
You should rebase this on nicks `-preview` PR, since that wasn't merged because there was no code that made use of it, and this needs that flag.

File: src/dmd/expressionsem.d (hunk 3 above)
Comment (UplinkCoder) [reply src/dmd/expressionsem.d]: I've put it into it's own function nyow.

File: test/fail_compilation/fail_shared.d [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+fail_compilation/fail_shared.d(16): Error: Trying to Access shared state `x`
+fail_compilation/fail_shared.d(17): Error: Trying to Access shared state `x`
+fail_compilation/fail_shared.d(22): Error: Trying to Access shared state `x`
+---
+*/
+
+
+int f1(shared int x)
+{
+    x += 7; // Error: Trying to accsess shared state x
```
Comment (TurkeyMan) [inline test/fail_compilation/fail_shared.d]: I think the error should say "Invalid access to shared data x"
'Trying' feels like an intent rather than an error.

File: src/dmd/expressionsem.d [hunk 8]
```diff
@@ -4011,8 +4032,45 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         return exp.expressionSemantic(sc);
     }
 
+    void checkAccessSharedCall(CallExp exp)
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]: No description.

File: src/dmd/expressionsem.d [hunk 9]
```diff
@@ -4011,8 +4032,45 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                    }
+                }
+            }
+        }
+    }
+
     override void visit(CallExp exp)
     {
+        // @@@SHARED@@@
+        // we need to set allow shared for overload resolution
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]: why should this affect overload resolution?

File: src/dmd/expressionsem.d [hunk 10]
```diff
@@ -6798,19 +6856,28 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             return;
         }
 
-        // for static alias this: https://issues.dlang.org/show_bug.cgi?id=17684
-        if (exp.e1.op == TOK.type)
-            exp.e1 = resolveAliasThis(sc, exp.e1);
-
-        auto e1x = resolveProperties(sc, exp.e1);
-        if (e1x.op == TOK.error)
+        // shared allowed for resolving this
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]: I have no idea what this means.

File: src/dmd/dscope.d [hunk 11]
```diff
@@ -57,6 +57,7 @@ enum SCOPE
     ignoresymbolvisibility    = 0x0200,   /// ignore symbol visibility
                                           /// https://issues.dlang.org/show_bug.cgi?id=15907
     onlysafeaccess = 0x0400,  /// unsafe access is not allowed for @safe code
+    nosharedcheck  = 0x0800, /// allow shared acsses for the sub-expressions
```
Comment (WalterBright) [inline src/dmd/dscope.d]: 1. I don't care for flags that negate something when on. 
2. how is this related to `restrictiveshare`?
3. fix spelling

File: src/dmd/dscope.d [hunk 12]
```diff
@@ -273,6 +274,19 @@ struct Scope
     }
 
 
+    extern (C++) Scope* startRelaxShared()
```
Comment (WalterBright) [inline src/dmd/dscope.d]: 1. No description
2. I have no idea what "relaxed shared" means

File: src/dmd/dscope.d [hunk 13]
```diff
@@ -273,6 +274,19 @@ struct Scope
...
+    {
+        Scope* sc = this.push();
+        sc.flags = this.flags | SCOPE.nosharedcheck;
```
Comment (WalterBright) [inline src/dmd/dscope.d]: Why is the function named "RelaxShared" and the associated flag is "nosharedcheck" which bears no obvious nameing relationship?

File: test/fail_compilation/fail_shared.d [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    x = 22; // Error: Trying to accsess shared state x
+    return x; // Error: Trying to accsess shared state x
+}
+
+int sync(shared int *x, int y)
+{
+    return x = y; // Error: Trying to accsess shared state x
```
Comment (WalterBright) [inline test/fail_compilation/fail_shared.d]: integers cannot be assigned to pointers anyway

File: test/fail_compilation/fail_shared.d (hunk 7 above)
Comment (WalterBright) [reply test/fail_compilation/fail_shared.d]: Spelling errors

File: src/dmd/cli.d [hunk 15]
```diff
@@ -721,6 +721,8 @@ dmd -cov -unittest myprog.d
             "destruct fields of partially constructed objects"),
         Feature("rvaluerefparam", "rvalueRefParam",
             "enable rvalue arguments to ref parameters"),
+        Feature("restrictiveshared", "restrictiveshared",
+            "implement a more restrictive shared", false),
```
Comment (WalterBright) [inline src/dmd/cli.d]: "more restrictive shared" has no particular meaning

File: src/dmd/expressionsem.d [hunk 16]
```diff
@@ -4011,8 +4032,45 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         return exp.expressionSemantic(sc);
     }
 
+    void checkAccessSharedCall(CallExp exp)
+    {
+        assert (exp.f);
+        auto params = exp.f.getParameterList().parameters;
+        if (exp.arguments && exp.arguments.dim && params && params.dim)
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d]: Please write this function as 
```
if (!(exp.arguments && exp.arguments.dim && params && params.dim)
    return;
foreach (i, arg;*exp.arguments)
{
    if (arg.op != TOK.variable)
        continue;
    ...
}

```

**Comment by thewilsonator** [comment]
@UplinkCoder please take a look at https://github.com/dlang/druntime/pull/2679#discussion_r305091533

File: src/dmd/cli.d (hunk 15 above)
Comment (stefan-koch-sociomantic) [reply src/dmd/cli.d]: `disallow reading and writing shared state` then?

File: test/fail_compilation/fail_shared.d (hunk 14 above)
Comment (stefan-koch-sociomantic) [reply test/fail_compilation/fail_shared.d]: auto-deref.

Comment (stefan-koch-sociomantic) [reply test/fail_compilation/fail_shared.d]: if (e2.type == e1.type.nextof) *e1 = e2

File: src/dmd/cli.d (hunk 15 above)
Comment (TurkeyMan) [reply src/dmd/cli.d]: Personally, I'd go with 'data' or 'values', in my melon 'state' means something a little more specific; it evokes emotions of state machines or perhaps the 'state' of whether something is shared or not, like "reading or writing the shared-ness of the thing"... I mean, it's obvious that's not what this error means, it's just that I would lean away from using the word 'state' for those reasons. Is there precedent elsewhere?

File: test/fail_compilation/fail_shared.d (hunk 14 above)
Comment (TurkeyMan) [reply test/fail_compilation/fail_shared.d]: > auto-deref.

You said that the other day, but I don't get it...
```D
    int y = 10;
    int* x;
    x = y;
```
```
error : cannot implicitly convert expression `y` of type `int` to `int*`
```

File: src/dmd/expressionsem.d (hunk 9 above)
Comment (UplinkCoder) [reply src/dmd/expressionsem.d]: why indeed :)
because for some reason overload-resolution needs to resolve properties.

File: src/dmd/expressionsem.d [hunk 17]
```diff
@@ -4011,8 +4032,50 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         return exp.expressionSemantic(sc);
     }
 
+    void checkAccessSharedCall(CallExp exp)
+    {
+        assert (exp.f);
+        auto params = exp.f.getParameterList().parameters;
+        if (!exp.arguments || !exp.arguments.dim || !params || !params.dim)
+            return ;
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d]: space

File: src/dmd/expressionsem.d [hunk 18]
```diff
@@ -4011,8 +4032,50 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+
+        {
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d]: remove and unindent the following

**Comment by RazvanN7** [comment]
@UplinkCoder It seems like Walter has taken the matter into his hands [1]. Maybe you can take a look at his implementation and see which one is better.

[1] https://github.com/dlang/dmd/pull/10209
